x-pack/solutions/observability/plugins/observability/server/services/dashboard_scorer.ts

```diff
@@ -0,0 +1,639 @@
+
+      return {
+        panel,
+        fieldScore: fieldScoreResult.score,
+        textScore: textScoreResult.score,
+        contributionToOverall,
+        fieldContributions: fieldScoreResult.fieldContributions,
+        textContributions: textScoreResult.textContributions,
+      };
+    });
+  }
+
+  /**
+   * Calculates the final combined score for a single dashboard.
+   * This method orchestrates the different scoring components.
+   */
+  private calculateCombinedScore(dashboard: RankableDashboard): ScoredDashboard {
+    // 1. Calculate panel-level scores, filtering out panels with no score
+    const panelScores = this.calculatePanelScores(dashboard).filter(
+      (score) => score.contributionToOverall > 0
+    );
+
+    // 2. Calculate the core scores for field and text matching
+    // We still keep the original dashboard-level field score for backwards compatibility
+    const fieldScoreResult = this.calculateFieldScore(dashboard);
+    const fieldScore = fieldScoreResult.score;
+
+    // For text similarity, we can use a weighted average of panel text scores and dashboard title
+    const textScoreResult = this.calculateTextSimilarity(dashboard);
+    const dashboardTitleScore = textScoreResult.score;
+
+    // Average text scores from panels
+    const avgPanelTextScore =
+      panelScores.length > 0
+        ? panelScores.reduce((sum, panel) => sum + panel.textScore, 0) / panelScores.length
+        : 0;
+
+    // Combined text score with more weight on dashboard title
+    const textScore = dashboardTitleScore * 0.6 + avgPanelTextScore * 0.4;
+
+    // 3. Calculate bonuses and penalties
+    const recencyBonus = this.calculateRecencyBonus(dashboard);
+    const defaultPenalty = this.getDefaultDashboardPenalty(dashboard);
+
+    // 4. Combine scores using a weighted average
+    // The base score is a weighted sum of the content-based scores.
+    const baseScore = this.W_FIELD * fieldScore + this.W_TEXT * textScore;
+
+    // 5. Apply recency bonus and default penalty
+    // Recency bonus is added to the weighted score.
+    // Default penalty is applied multiplicatively to reduce the score.
+    let finalScore = (baseScore + this.W_RECENCY * recencyBonus) * defaultPenalty;
+
+    // Ensure score is capped between 0 and 1
+    finalScore = Math.max(0, Math.min(finalScore, 1));
+
+    return {
+      dashboard,
+      score: finalScore,
+      scoreBreakdown: {
+        fieldScore,
+        textScore,
+        recencyBonus,
+        defaultPenalty,
+        finalScore,
+        fieldContributions: fieldScoreResult.fieldContributions,
+        textContributions: textScoreResult.textContributions,
+        panelScores,
+      },
+    };
+  }
+
+  /**
+   * Calculates a score based on shared fields, factoring in frequency.
+   * This is more nuanced than a simple intersection/union.
+   * We'll use cosine similarity on term frequency vectors.
+   */
+  private calculateFieldScore(dashboard: RankableDashboard): {
```

Comment:
is it td-idf algo?

Reply:
I can't remember all the details of tf-idf, but its similar